### DISCUSSION TITLE
KLIFF 1.0

### DISCUSSION THREAD

**Discussion by ipcamit** [ipcamit/kliff#2] [Ideas]
I created this on saturday as Project, but I guess thats not visible to you guys so starting it as a discussion.

@mjwen and @yonatank93

As per our discussion I took the liberty to come up with a basic outline for KLIFF 1.0. Let me know of your thought and suggestions. Once we have more concrete workplan, we can start incorporating things in main KLIFF.

### Basic introduction of KLIFF 1.0 structure and plans.
Proposed major modules:

```python
kliff.uq             # Uncertainty Quantification tools
kliff.transformation # Transformation classes for Configurations and Properties
kliff.ml             # ML modules specifically TrainingWheels
kliff.optimizer      # OptimizerScipy and OptimizerTorch
kliff.data           # Dataset, Configuration, Parameters etc
kliff.models         # Developed models, universal ML models etc
kliff.utils
kliff.neighbor       # ? not needed ?
kliff.legacy         # Old stuff for backward compatibility

```

Details:
1. `kliff.uq` Would be focus of most development. It would have a new base class called `kliff.uq.Sampler`, which will implement `Bootstrap` and `MCMC` sampling, and also pave way for future additions. Most notably among future additions would be `HMC` for ML uncertainty models.
2. `kliff.transformation` is the second major new contribution. It is supposed to implement functions/classes that would transform various members of `kliff.data` . I would need some more work to suggest exact form and function of it. Idea is that it can be divided in three broad category
	1. `kliff.transformation.CoordinateTransformation`:Takes in `Configuration` object and gives out transformed data like, Descriptors, Graphs, Cell transforms etc. This unifies the concept of transformation and descriptors.
	2. `kliff.transformation.PropertyTransformation`: Takes in `Dataset` object, traverses through all the properties and apply global transforms. This would include weight for loss functions as well. Example: `AffineTransform`
	3. `kliff.transformation.ParameterTransformation`: For aiding UQ, it can be used to transform the `Parameter` object. Example: `LogTransform`
3. `kliff.ml` Would contain wrappers for making descriptors and graphs compatible with `torch.autograd.grad` API and contain the `TrainingWheels`
 module. It is mostly done need little to no development.
4. `kliff.optimizer `, module for  `ScipyOptimizer`, `TorchOptimizer`
5. `kliff.data`,  will contain the `Configuration` object , `Dataset` object, `Parameters` object. Out of which Configuration and Dataset are already there. Need to polish the parameter object a bit.
6. `kliff.models` and  `kliff.utils` no special modification required. Mostly there. Except NeighborList might become part of utils.

7. `kliff.legacy`, for containing old modules, like calculator object.

**Comment by yonatank93** [comment]
I am guessing that `kliff.uq` will only contain, with the language I used in my presentation, generic UQ methods.

If we have `kliff.uq.Sampler`, can the user still access `Bootstrap`, `MCMC`, and `HMC` directly? I think there are some users who would prefer to access these methods directly. And as a record as well, there are other sampling methods that we can implement in the future, such as some Monte Carlo-type of methods and some modifications of MCMC and HMC.

I also found that there are other UQ methods that don't use any sampling, such as Fisher information analysis and polynomial chaos expansion (PCE). If we have `kliff.uq.Sampler`, is it going to overshadow other, non-sampling methods? That is, will it sound like we encourage people to use the sampling methods more than, say, FIM? Any thoughts on this?

On a different note, I think that we need to move the FIM calculation to `kliff.uq`, but I can do this in the future.

**Reply by ipcamit** [reply]
At present I was not moving FIM, as it fits fine in `analyzers` as well. But if needed we can move it in `kliff.uq`. `Bootstrap` and `MCMC` can be implemented as separate class if needed, but `Sampler` would be one unified wrapper for them (Although it would be nice if they share common design for ease of maintenance).

That way for lower level access you can use them directly, but for basic use cases you can have `Sampler`. Same goes for HMC, though currently HMC would be used for ML models mainly.

This also gives us leeway for adding other methods in future like VI.

what do you think?

**Reply by mjwen** [reply]
I agree. In the `Sampler` class, there could be a `sampler` attribute that points to the real sampling methods, be it bootstrap or MCMC.

**Reply by mjwen** [reply]
I am debating whether we want to keep the `analyzer` submodule. The name suggests that everything inside it is for analysis. But in reality, many stuff can be used in multiple places. FIM is one example, and detailed MAE/RMSE is another example, which provides analysis but can also be used as loss functions.  

So, we might want to regroup stuff according to what they are fundamentally doing, not what we are thinking they can be used for.

**Reply by yonatank93** [reply]
I would vote if we could implement `Bootstrap` and `MCMC` as separate classes and make them available for users if they want to use them directly. In that case, then `kliff.uq.Sampler` would be a wrapper class over those classes. Also, are you thinking of including HMC in `Sampler`?

**Reply by yonatank93** [reply]
There are other methods that Mark and I are already thinking of adding to KLIFF, for example, some Monte Carlo-type methods. I think it would be better to implement these sampling methods in separate classes and, again, `Sampler` acts as a wrapper. I think this will make it more modular.

**Reply by mjwen** [reply]
Wait, I think implementing`Sampler` as a base class defining the API (e.g. abstract methods), and then `Bootstrap`, `MCMC` etc subclass it and implementing concrete methods would be easier from the maintaining perspective. Is there anything preventing us defining an API unifying bootstrap, mcmc, and some other MC-type methods?

**Reply by yonatank93** [reply]
Are you thinking of something like the `Weight` class, where if we want to use a different way to compute weights, we define a class that inherits from the base `Weight` class?

Currently, each sampler needs different additional arguments. For the current methods, they need the loss function, however, I cannot guarantee that other methods would also use the loss function. But for the UQ methods I know, we are good. And for general workflow, I think MCMC, bootstrapping, HMC, and other sampling methods have the same pattern: we instantiate the sampler, then run the sampling. For the current bootstrapping, there is an extra step, but we can also merge that into `run` method.

However, there is something to consider for MCMC. After we run the sampling, we still need to do some post-processing to get the samples, which, due to some reason, is usually done manually.

I might want to try playing around and try to follow this suggestion to give a better response.

**Reply by ipcamit** [reply]
I am fine either way. The design decision is basically usability vs maintainability. Having MCMC and Bootstrap inherit from a base `Sample` is easy to maintain and contribute to, while `Sample(method="Bootstrap")` might be more intuitive to the user. But again I am fine either way.

My "inspiration" was basically this: https://turing.ml/dev/docs/using-turing/guide have a look at the `sample` method. Regarding post-processing etc, I think we would need some sort of common framework for such modifications. One I was thinking was that every required class will have a pre and post processing function that you can implement as needed. But will discuss it more after having few examples.

**Comment by mjwen** [comment]
Thanks @ipcamit for documenting it. This looks great! 

For `kliff.uq`, an ML method (besides further addition of `HMC`) we need to implement is the dropout approach I've discussed on Friday. It is currently implemented in the NN model class, but we need to migrate it to be under this umbrella. Both Yonatan and Zeren are using the dropout stuff now and we need to support it out of the box. 

For `kliff.transformation.PropertyTransformation`, I'm still debating whether it should take `Configuration` or `Dataset` as input. I have arguments for both. When using `Configuration`, it is more clean and would have a similar interface as `kliff.transformation. CoordinateTransformation`; the flip side is that it may need dataset statistics as input (e.g. mean energy of dataset). When using `Dataset`, it is bulky (incorporating a lot of stuff) which I don't like. Probably, we can implement both: the one taking `Dataset` as input simply wraps around the one taking `Configuration`. 

Also, I have long been planning to redo the doc system. The current one includes a lot of unnecessary files and it is hard to manage. Below are the plans: 
- replacing `.rst` files with `MyST Markdown`, which provides all the necessary functionality as in .rst, and the biggest benefit is that people are more familiar with markdown, making it easier for creating docs   
- using [jupyterbook](https://jupyterbook.org/en/stable/intro.html) as the build system for docs 
- replacing the current tutorial examples in .py files with jupyter notebooks, and providing binder/colab links to run them directly
- adding more developer/contributing guide 

We will probably redo the docs after everything else is finalized. But when you create new docs (inline docstring or standalone .md files), please use the [MyST Markdown](https://myst-tools.org/docs/spec).

**Reply by ipcamit** [reply]
Will the dropout uncertainty use the DUNN driver or are proposing a design from scratch. What `kliff.uq` currently does is measure parametric uncertainty. How do you propose we incorporate dropout stuff? Should we try and implement it using the new model driver? It will not be as performant as the DUNN driver though.

**Reply by mjwen** [reply]
The current DUNN driver is very specific -- with a simple MLP on symmetry function descriptors. I don't think we want to make it general. Instead, we want the general stuff from the new driver. I guess no modification to the driver is needed, right? 

I don't think we want to limit `kliff.uq` to the parametric uncertainty of physics-based models, right? In fact, bootstraping can be applied to an NN model as well (I believe we are refactoring the code to enable this). I see that dropout is very special, but we need to provide a user-friendly implementation and place it somewhere. If `kliff.uq` is not a good place, we might want to provide a `DropoutNNModel` in `kliff.models`.

**Reply by yonatank93** [reply]
Just want to confirm that `Bootstrap` supports NN model, and, as @mjwen said, this is done by splitting it into 2 classes: one for physics-based and one for NN model.

**Reply by mjwen** [reply]
Ah, right, I recall that. Thanks for the reminder!

I guess in the refactoring, we want to try unifying some of the API of physics models and ML models (if possible), so we do not need to use many `if-else` statements, which makes long-term maintenance a pain.

**Reply by ipcamit** [reply]
I would suggest that `kliff.models` might be best place for such specialized models that we will provide pre-trained and people can do transfer learning etc over. I think we can make it work with current model driver, but if needed we can look in more specialized driver as well. 

Also as `Bootstrap` only needs the models trained, i dont think it will need too many if-else statements etc. For example we can simple pass an appropriate `Optimizer` instance to the `Bootstrap` class, and it can minimize the ensemble parameters. That way Bootstrap is decoupled from model optimization.

**Reply by mjwen** [reply]
Agreed regarding dropout in  `kliff.models`

If I remember correctly, physics-based and ML models are largely treated differently in the current implementation of bootstrapping (correct me @yonatank93). In the new implementation, we'd like to reduce such disparity as much as possible. And I think we can just as you described @ipcamit.

**Reply by yonatank93** [reply]
Yes, currently the if-else statement in `Bootstrap` (or why I splitters up into 2 classes) is because what we now call compute arguments are treated differently between the physics-based and NN models. In general, if we can resolve this difference, maybe by making the names of the properties the same, we can just have 1 `Bootstrap` class for both models.

**Reply by ipcamit** [reply]
Yes. That should be possible after the refactoring of parameter class, in fact let me put that as the first to-do task. It is nearly done and need some polish.
